### PR TITLE
Fix WeakHash32 collision and ColumnUnique Field insertion issue on s390x

### PR DESCRIPTION
**Pull request by HarryLeeIBM**
The PR addressed the issue that field can not be inserted correctly into ColumnUnique when Field contains integer which is shorter than 64 bits on s390x.

The root of cause is that when inserting short(non-64 bit) integer Field into ColumnUnique, the code assumes the 64 bit of integer in Field is little-endian. On big-endian machine, the starting position of short integer is not at beginning.
The fix to calculate correct starting position based on the Field type.

### Changelog category (leave one):

- Build Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
...
Fixed ColumnUnique short integer Field insertion issue on s390x

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Review by vdimir — APPROVED** [review]

File: src/Common/HashTable/Hash.h [hunk 1]
```diff
@@ -48,15 +48,22 @@ inline DB::UInt64 intHash64(DB::UInt64 x)
...
 {
 #ifdef __SSE4_2__
     return _mm_crc32_u64(-1ULL, x);
 #elif defined(__aarch64__) && defined(__ARM_FEATURE_CRC32)
     return __crc32cd(-1U, x);
 #else
-    /// On other platforms we do not have CRC32. NOTE This can be confusing.
-    return intHash64(x);
+    /// On other platforms we use zlib CRC
+    return zlibCRC32(-1ULL, x);
```
Comment (alexey-milovidov) [inline src/Common/HashTable/Hash.h:66]: I don't think this is a good option.

CRC-32 has lower quality than intHash64,
and it has lower performance on hardware platforms without the corresponding instructions.

**Review by alexey-milovidov — CHANGES_REQUESTED** [review]
You can get more than zero collisions, but not significantly more.
Please check why exactly you get this number of collisions.

This PR shouldn't be merged as is.

File: src/Common/HashTable/Hash.h (hunk 1 above)
Comment (HarryLeeIBM) [reply src/Common/HashTable/Hash.h]: @alexey-milovidov, I rollbacked the change of Hash.h and updates thresholds in weakhash32 unit test for s390x, please review again. Thanks.

**Comment by HarryLeeIBM** [comment]
Closed the PR because of conflicts, will re-create PR.

**Comment by uweigand** [comment]
It appears to me the problem might lie not so much with the hash function, but with how it is used, in particular in `updateWeakHash32`:
```
        /// If string size is not divisible by 8.
        /// Lets' assume the string was 'abcdefghXYZ', so it's tail is 'XYZ'.
        DB::UInt8 tail_size = end - pos;
        /// Load tailing 8 bytes. Word is 'defghXYZ'.
        auto word = unalignedLoad<UInt64>(end - 8);
        /// Prepare mask which will set other 5 bytes to 0. It is 0xFFFFFFFFFFFFFFFF << 5 = 0xFFFFFF0000000000.
        /// word & mask = '\0\0\0\0\0XYZ' (bytes are reversed because of little ending)
        word &= (~UInt64(0)) << DB::UInt8(8 * (8 - tail_size));
        /// Use least byte to store tail length.
        word |= tail_size;
        /// Now word is '\3\0\0\0\0XYZ'
        updated_value = intHashCRC32(word, updated_value);
```
This computation assumes little-endian byte order.  On a big-endian system, I think it may zero out the wrong half of the loaded bytes, which may very well result in many of the actual input bytes to be zeroed out -- that could lead to many of the hashes being identical (certainly in the case of short input strings).

**Comment by HarryLeeIBM** [comment]
@uweigand , thanks for your comment. I agree this code can affect hash quality on big-endian machine, however the issue happens on ppc64le(little endian) too. It can happen on x86 too if we force it using  intHash64 by hacking the code. So I believe it is a more complicated issue than that.

**Pull request closed without merging**